### ISSUE TITLE
Fulcrum says - Incompatible version of OpenSSL

### ISSUE THREAD

**Issue by calligrl** [cculianu/Fulcrum#126]
I'm following the Ministry of Nodes bitcoin node setup on youtube and get this error when trying to setup fulcrum to start as a service:

After doing 'enable' and 'start'.

satoshi@localhost:~> sudo systemctl status fulcrum.service
× fulcrum.service - Fulcrum
     Loaded: loaded (/etc/systemd/system/fulcrum.service; enabled; vendor preset: disabled)
     Active: failed (Result: exit-code) since Tue 2022-08-02 01:59:16 BST; 4s ago
    Process: 7442 ExecStart=/home/satoshi/fulcrum/Fulcrum /home/satoshi/fulcrum/fulcrum.conf (code=exited, status=1/FAILURE)
   Main PID: 7442 (code=exited, status=1/FAILURE)

Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.267] [Qt Warning] QSslSocket: cannot resolve SSLv23_server_method (:0, )
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.267] [Qt Warning] QSslSocket: cannot resolve X509_STORE_CTX_get_chain (:0, )
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.267] [Qt Warning] QSslSocket: cannot resolve OPENSSL_add_all_algorithms_noconf (:0, )
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.267] [Qt Warning] QSslSocket: cannot resolve OPENSSL_add_all_algorithms_conf (:0, )
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.267] [Qt Warning] QSslSocket: cannot resolve SSLeay (:0, )
Aug 02 01:59:16 localhost.localdomain systemd[1]: fulcrum.service: Main process exited, code=exited, status=1/FAILURE
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.267] [Qt Warning] Incompatible version of OpenSSL (:0, )
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.269] SSL support is not compiled and/or linked to this version. Cannot proceed with SSL support. Sorry!
Aug 02 01:59:16 localhost.localdomain Fulcrum[7442]: [2022-08-02 01:59:16.269] Use the -h option to show help.
Aug 02 01:59:16 localhost.localdomain systemd[1]: fulcrum.service: Failed with result 'exit-code'.
satoshi@localhost:~>

I'm using openSUSE Leap 15.4 and have installed:

openssl
openssl-1-1
openssl-3

Which version is it that fulcrum needs?

**Comment by cculianu** [comment]
Well it normally needs 1.1.  But if you download the old `ub16` version which I provide (intended for ubuntu16), that may be the problem. That older one works with openssl 1.0 only I believe.   Try the [linux version that doesn't have the `ub16` suffix](https://github.com/cculianu/Fulcrum/releases/download/v1.7.0/Fulcrum-1.7.0-arm64-linux.tar.gz) which should work on with openssl 1.1.

Let me know if that was it or if the problem persists.

**Comment by calligrl** [comment]
I'm running openSUSE 15.4 and this is the version of Fulcrum I installed:

https://github.com/cculianu/Fulcrum/releases/download/v1.7.0/Fulcrum-1.7.0-x86_64-linux.tar.gz

**Comment by cculianu** [comment]
Hmm. Well you may have to compile from source then... seems like the static Qt linked into the app doesn't like the version of openSSL that is installed.. it's probably finding openssl-3 first. and not finding openssl-1.1 for some reason.  You can try messing with LD_LIBRARY_PATH env var to get it to search for the dir where openssl-1.1 lives first.  

If that doesn't work you may have to just build Fulcrum from source...

**Comment by calligrl** [comment]
Thanks. I'll look into it.

**Comment by mbrochh** [comment]
I'm stuck here as well.

I'm running:

* Ubuntu 22.04.1 LTS
* OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
* Fulcrum downloaded from https://github.com/cculianu/Fulcrum/releases/download/v1.8.0/Fulcrum-1.8.0-x86_64-linux.tar.gz

I get the following error:

```
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: cannot resolve EVP_PKEY_param_check (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: cannot resolve SSL_CTX_set_ciphersuites (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: cannot resolve SSL_set_psk_use_session_callback (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: cannot resolve SSL_SESSION_is_resumable (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: OpenSSL >= 1.1.1 is required; OpenSSL 1.1.0l  10 Sep 2019 was found instead (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: OpenSSL >= 1.1.1 is required; OpenSSL 1.1.0l  10 Sep 2019 was found instead (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: OpenSSL >= 1.1.1 is required; OpenSSL 1.1.0l  10 Sep 2019 was found instead (:0, )
[2022-09-17 03:41:16.566] [Qt Warning] QSslSocket: OpenSSL >= 1.1.1 is required; OpenSSL 1.1.0l  10 Sep 2019 was found instead (:0, )
[2022-09-17 03:41:16.567] [Qt Warning] QSslSocket: OpenSSL >= 1.1.1 is required; OpenSSL 1.1.0l  10 Sep 2019 was found instead (:0, )
[2022-09-17 03:41:16.567] SSL support is not compiled and/or linked to this version. Cannot proceed with SSL support. Sorry!
```

**Comment by mbrochh** [comment]
NOTE: This works for me:

```
wget http://nz2.archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2.16_amd64.deb
sudo dpkg -i libssl1.1_1.1.1f-1ubuntu2.16_amd64.deb
```

but I'm not sure what the implications of this are. Does this just install libssl1.1.1 alongside the latest version, or does this downgrade the entire system to 1.1.1, which surely would not be a good idea?

**Comment by cculianu** [comment]
You know what? This is causing everybody such a headache. I think I will investigate statically linking the appropriate openssl into the Linux Fulcrum static binary.  I did this for the Windows build already (given how Windows typically lacks openssl at all).  Bitcoin Core and Bitcoin Cash Node already do this anyway in their static builds.. I will do it for linux Fulcrum I think.  Too many people are having trouble and there is no clear path for me on how to resolve this seeing as how Qt5 really requires openssl 1.1.1 (which is really what Fulcrum was well tested against, although Qt6 apparently works too).

**Comment by mbrochh** [comment]
That sounds amazing! Ping me here when you got a new release. I'll test it and if it works, I'll send you 0.01 BTC.

**Comment by cculianu** [comment]
@mbrochh I made some binaries with openssl statically linked-in that appears to run much better across many distros because it no longer depends on whatever quirky openssl install happens to be present. It just statically links the exact openssl 1.1.1 lib into Fulcrum that it needs (at the expense of bloating the Fulcrum binary a little bit.. heh).

Here is a link to some Fulcrum test binaries for x86_64 and for arm64 that are linked this way: https://c3-soft.com/downloads/BitcoinCash/Fulcrum/testing/openssl-static/

Would you or anyone in this thread be able to try the appropriate one out for yourself and let me know if things are "better"?  In my testing no warnings appear and everything seems peachy.  

If I get confirmation on it I can prepare a 1.8.1 release that only differs in this 1 aspect from 1.8.0 (has openssl statically-linked).

**Comment by cculianu** [comment]
Oops, my commit auto-closed this.  I will leave this open until I do a release with this and it's confirmed it works for you guys..

**Comment by mbrochh** [comment]
@cculianu I will try to test this later today.

**Comment by cculianu** [comment]
Ok -- no rush.

I'm pretty sure it works 100% now I have been testing and re-testing it on vanilla virgin Linux installs for ubuntu22, ubuntu18, ubuntu16, etc, both on AMD64 and ARM64. The static binary now does 100% incorporate a private, statically-built and linked openssl 1.1.1o.  This should mean that the Fulcrum binary should more reliably work on any system since it no longer will depend on the (fragile) setup of whatever openssl version the system happened to have. (which was the root cause of this issue).

I'm preparing a release of Fulcrum v1.8.1 now that incorporates these changes and I will close this issue when that's complete.

**Comment by mbrochh** [comment]
@cculianu  alright. I wiped my nodebox, reinstalled Ubuntu server and UNFORTUNATELY, it all seemed to work this time, even without your new binary.

When I did `./Fulcrum --version` it showed OpenSSL 3 as the used version.

I replaced the binary with the one you provided and now it shows `ssl: OpenSSL 1.1.1  11 Sep 2018` and things seem to work as well.

So all in all, looks good to me.

**Comment by cculianu** [comment]
Welp, that's good! Ha ha. Anyway I did do a release, since others have had this problem too (See issue #132 ).  It's finally solved in release v1.8.1, which was just published.
